File: frontend/src/types/insights.ts
export type AggregateMetrics = {
  matches: number;
  wins: number;
  losses: number;
  draws: number;
  win_rate: number | null;
  avg_acs: number | null;
  avg_adr: number | null;
  avg_kda: number | null;
  avg_hs_percent: number | null;
  avg_rr_change: number | null;
};

export type BreakdownEntry = AggregateMetrics & {
  label: string;
};

export type MetricDelta = {
  metric: string;
  recent: number | null;
  baseline: number | null;
  delta: number | null;
  direction: "up" | "down" | "flat";
};

export type TrendSummary = {
  recent_window: number;
  baseline_window: number;
  comparisons: MetricDelta[];
};

export type StreakSummary = {
  current_streak_type: string;
  current_streak_length: number;
  longest_win_streak: number;
  longest_loss_streak: number;
};

export type VolatilitySummary = {
  acs_std_dev: number | null;
  rr_std_dev: number | null;
  result_switch_rate: number | null;
  level: string;
};

export type InsightsResponse = {
  recent_form: AggregateMetrics;
  baseline_form: AggregateMetrics;
  trend_summary: TrendSummary;
  streaks: StreakSummary;
  volatility: VolatilitySummary;
  map_breakdowns: BreakdownEntry[];
  agent_breakdowns: BreakdownEntry[];
  role_breakdowns: BreakdownEntry[];
};

File: backend/app/schemas/insights.py
from pydantic import BaseModel, Field


class AggregateMetrics(BaseModel):
    matches: int = 0
    wins: int = 0
    losses: int = 0
    draws: int = 0
    win_rate: float | None = None
    avg_acs: float | None = None
    avg_adr: float | None = None
    avg_kda: float | None = None
    avg_hs_percent: float | None = None
    avg_rr_change: float | None = None


class BreakdownEntry(AggregateMetrics):
    label: str


class MetricDelta(BaseModel):
    metric: str
    recent: float | None = None
    baseline: float | None = None
    delta: float | None = None
    direction: str = "flat"


class TrendSummary(BaseModel):
    recent_window: int
    baseline_window: int
    comparisons: list[MetricDelta] = Field(default_factory=list)


class StreakSummary(BaseModel):
    current_streak_type: str = "none"
    current_streak_length: int = 0
    longest_win_streak: int = 0
    longest_loss_streak: int = 0


class VolatilitySummary(BaseModel):
    acs_std_dev: float | None = None
    rr_std_dev: float | None = None
    result_switch_rate: float | None = None
    level: str = "insufficient_data"


class InsightsSummary(BaseModel):
    recent_form: AggregateMetrics
    baseline_form: AggregateMetrics
    trend_summary: TrendSummary
    streaks: StreakSummary
    volatility: VolatilitySummary
    map_breakdowns: list[BreakdownEntry] = Field(default_factory=list)
    agent_breakdowns: list[BreakdownEntry] = Field(default_factory=list)
    role_breakdowns: list[BreakdownEntry] = Field(default_factory=list)
